src/pc/protocols/tcpip_host.cpp

```diff
@@ -1163,6 +1233,16 @@ int tcpipPlatformBootFirmware(const deviceDesc_t* deviceDesc, const char* firmwa
     return -1;
 }
```

Comment:
And - move the thread & vars into func scope or struct. should look something like this in the end:

```
tcpipOrLocalShdmemServer(*protocol) {
    if(protocol == nullptr) return;

    std::mutex mtx;
    std::condition_variable cv;
    bool isTcpIpThreadFinished = false;
    bool isShdmemThreadFinished = false;
    int retTcpIp = -1, retShdmem = -1;
    void *fdTcpIp = nullptr, *fdShdmem = nullptr;

    auto thread_tcpip = std::thread([&mtx, &cv, &isTcpIpThreadFinished, &retTcpIp, &fdTcpIp](){
        auto ret = tcpipPlatformServer(..., &fdTcpIp);
        {
            std::unique_lock<std::mutex> l(mtx);
            isTcpIpThreadFinished = true;
            retTcpIp = ret;
        }
        cv.notify_one();
    });  

    auto thread_shdmem = std::thread([&mtx, &cv, &isShdmemThreadFinished, &retSdhmem, &fdShdmem](){
        auto ret = shdmemPlatformServer(..., &fdShdmem);
        {
            std::unique_lock<std::mutex> l(mtx);
            isShdmemThreadFinished = true;
            retSdhmem = ret;
        }
        cv.notify_one();
    });  

    {
        std::unique_lock<std::mutex> lock(mtx);
        cv.wait(lock, [&isShdmemThreadFinished, &isTcpIpThreadFinished]{ return isShdmemThreadFinished || isTcpIpThreadFinished; });

        // As soon as either one finishes, the other can be cleaned
        // Use signals, as "accept" cannot be unblocked by "close"ing the underlying socket
        if(!isShdmemThreadFinished) pthread_kill(thread_shdmem.native(), SIGINT);
        if(!isTcpIpThreadFinished) pthread_kill(thread_tcpip.native(), SIGINT);
    }

    // Wait for both threads to wrap up
    if(thread_tcpip.joinable()) thread_tcpip.join();
    if(thread_shdmem.joinable()) thread_shdmem.join();

    // Asign the final protocol (once both threads finalize)
    if(retTcpIp == 0) *protocol = X_LINK_TCP_IP;
    if(retShdmem == 0) *protocol = X_LINK_LOCAL_SHDMEM;

    // if both connected, close TCP_IP
    if(retTcpIp == 0 && retShdmem == 0) {
        tcpipPlatformClose(fdTcpIp);
    }

    // None of the connections were successful
    if(retTcpIp || retShdmem) return -1;

    ...
}
```

That is if "signaling" works as expected with "accept". Otherwise you might have to switch to `select` in the servers, for which might be easiest to modify funcs with additional parameter with a fd that one can "unblock" (instead the pthread_kill calls)

Reply:
Signaling will work, but there's an issue: we don't clean up the socket instantiated with the socket() syscall. To stop the accept call, it is sufficent to use this procedure:
```
shutdown(sock, SHUT_RDWR);
#if defined(SO_LINGER)
const int set = 0;
setsockopt(sock, SOL_SOCKET, SO_LINGER, (const char*)&set, sizeof(set));
#endif
tcpip_close_socket(sock);
```

Since shutdown blocks any data transfer on the socket and setting SO_LINGER to zero means that when calling close() the socket will just be forced closed.
This has the added benefit of the thread ending gracefully and not needing a SIGINT.

Reply:
@TheMutta 

AFAIK (as per reading yesterday), `accept` cannot be unblocked with shutdown
We'd have to switch to `select`

If above works however, I'm up for that path - rather than signals

> That's the reason detach() was called, because the thread ends normally with no need of intervention.

lets join if thread is joinable - rather than detach (if thread actually finished as expected)

Reply:
The issue seems to be that signals cause issues, they don't work, while the shutdown(),setsockopt() and close() methods in succession do work. While shutdown() isn't the one closing the socket, it's instead the combination of setting SO_LINGER to 0 and then calling close() which forces an immediate closure of the socket, it's sort of a security measure to make sure the socket is neither sending or receiving data and is in the middle of something (if there's a pending connection request and the socket, for example, we don't want it to block in the middle, we can call shutdown and let it complete that request and then just close the socket).

> lets join if thread is joinable - rather than detach (if thread actually finished as expected)

Gotcha.

Reply:
Noted - lets create a duplicate of "tcpipServer" function for this, where we have this (mostly) same logic flow as the original function, but with these changes (to not cause potential issues for the regular TCP_IP flow)

Thanks!